### PR TITLE
first draft for RP2XXX docs

### PR DESCRIPTION
**Pull request by NitiKaur**
@mattytrentini @jimmo @dpgeorge   please review afd8ff5 . Also, I had this doubt regarding the troubleshooting section that I have mentioned on slack.

### CODE REVIEW

**Review by mattytrentini — COMMENTED** [review]
I've made a start on the review but will continue tomorrow - and we can discuss further at the meeting.

**Comment by NitiKaur** [comment]
@mattytrentini I have made the changes wherever it was suggested. @dpgeorge I have edited the quickref to include only the most relevant piece of information as well as written the three files in my own words as discussed over the meeting

File: docs/rp2/general.rst [hunk 1]
```diff
@@ -10,29 +10,30 @@ the RP2040.
 Technical specifications and SoC datasheets
 -------------------------------------------
 
-If you want a deatiled technical specification, then you can view the `datasheets <https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf>`_ 
+For a deatiled technical specification, then you can view the `datasheets <https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf>`_
```
Comment (mcauser) [inline docs/rp2/general.rst:13]: deatiled -> detailed

File: docs/rp2/general.rst [hunk 2]
```diff
@@ -10,29 +10,30 @@ the RP2040.
...
 
 # 133 MHz dual ARM Cortex-M0+ cores (overclockable to over 400 MHz)
-	* Each core has an integer divider peripheral, and two interpolators.
-# 264 KB SRAM in six independent banks
-# No internal Flash or EEPROM memory (after reset, the boot-loader loads firmware from either external flash memory or USB bus into internal SRAM)
-# QSPI bus controller, supporting up to 16 MB of external Flash memory
+	* Each core has an integer divider peripheral and two interpolators
+# 264KB SRAM in six independent banks
+# No internal Flash or EEPROM memory (after reset, the bootploader loads firmware from either the external flash memory or USB bus into internal SRAM)
+# QSPI bus controller, which supports upto 16 MB of external Flash memory
```
Comment (mcauser) [inline docs/rp2/general.rst:21]: upto -> up to

File: docs/rp2/general.rst [hunk 3]
```diff
@@ -10,29 +10,30 @@ the RP2040.
...
+The RP2040 microcontroller a 40 nm silicon in a 7X7mm QFN-56 SMD package.The key features include:
 
 # 133 MHz dual ARM Cortex-M0+ cores (overclockable to over 400 MHz)
-	* Each core has an integer divider peripheral, and two interpolators.
-# 264 KB SRAM in six independent banks
-# No internal Flash or EEPROM memory (after reset, the boot-loader loads firmware from either external flash memory or USB bus into internal SRAM)
-# QSPI bus controller, supporting up to 16 MB of external Flash memory
+	* Each core has an integer divider peripheral and two interpolators
+# 264KB SRAM in six independent banks
+# No internal Flash or EEPROM memory (after reset, the bootploader loads firmware from either the external flash memory or USB bus into internal SRAM)
```
Comment (mcauser) [inline docs/rp2/general.rst:20]: bootploader?

File: docs/esp32/quickref.rst [hunk 4]
```diff
@@ -603,3 +603,31 @@ the corresponding functions, or you can use the command-line client
 
 See the MicroPython forum for other community-supported alternatives
 to transfer files to an ESP32 board.
+
+Interrupts
+----------
+**Hardware Interrupts** - They are fired by hardware peripherals and sources.hey can be: Internal or External. An example of internal hardware interrupts maybe something like a hardware timer interrupt or WDT. An example of an external hardware interrupt is external GPIO pins interrupts
```
Comment (mcauser) [inline docs/esp32/quickref.rst:609]: sources.hey -> sources. They

File: docs/rp2/quickref.rst [hunk 5]
```diff
@@ -59,7 +59,16 @@ Use the :mod:`time <utime>` module::
 Timers
 ------
 
-How do they work?
+The system timer peripheral on RP2040 provides a global microsecond timebase for the system, and generates
+interrupts based on this timebase. It supports • A single 64-bit counter, incrementing once per microsecond;• This counter can be read from a pair of latching registers, for race-free reads over a 32-bit bus;• Four alarms: match on the lower 32 bits of counter, IRQ on match.The timer uses a one microsecond reference that is generated in the watchdog and derived from the refernce clock,which itself is usually connected to the crystal oscillator.The 64-bit counter effectively can not overflow (thousands of years at 1 MHz), so the system timer is completely monotonic in practice.
```
Comment (mcauser) [inline docs/rp2/quickref.rst:63]: A few `•` chars made it in there

File: docs/rp2/quickref.rst [hunk 6]
```diff
@@ -92,11 +101,21 @@ See :ref:`machine.UART <machine.UART>`. ::
     uart1.write('hello')  # write 5 bytes
     uart1.read(5)         # read up to 5 bytes
 
+.. note::
+ 	
+	REPL over UART is disabled by default.You can see the :ref:`rp2_intro` for details on how to enable REPL over UART.
```
Comment (mcauser) [inline docs/rp2/quickref.rst:106]: default.You -> default. You

File: docs/rp2/quickref.rst [hunk 7]
```diff
@@ -92,11 +101,21 @@ See :ref:`machine.UART <machine.UART>`. ::
...
 ----------------------------
 
-How does PWM work on the RPi RP2xxx?
+
+There are 8 independent channels each of which have two outputs making it 16 PWM channels in total which can be clocked from 7Hz to 125Mhz. 
+
+
+The channels can be clocked by external sources, using an edge-sensitive input mode for frequency measurement, and a level-sensitive input mode for duty cycle measurement.
+
+For more information on this topic you can see the PWM module in the rp2040 datasheet.
```
Comment (mcauser) [inline docs/rp2/quickref.rst:118]: Add another link to it?
`datasheet <https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf>`

File: docs/rp2/quickref.rst [hunk 8]
```diff
@@ -112,7 +131,9 @@ Use the ``machine.PWM`` class::
 ADC (analog to digital conversion)
 ----------------------------------
 
-How does the ADC module work?
+The ADC on RP2040 has a 12-bits resolution but with MicroPython it is transformed to a 16-bit number.
+ 
+RP2040 has five ADC channels in total, four of which are brought out to chip GPIOs: GP26, GP27, GP28 and GP29. On Raspberry Pi Pico, the first three of these are brought out to GPIO pins. The fourth is used to measure the VSYS voltage on the board.The ADC’s fifth input channel is connected to a temperature sensor built into RP2040.
```
Comment (mcauser) [inline docs/rp2/quickref.rst:136]: board.The -> board. The

File: docs/rp2/quickref.rst [hunk 9]
```diff
@@ -208,7 +229,7 @@ See :ref:`machine.RTC <machine.RTC>` ::
 WDT (Watchdog timer)
 --------------------
 
-Is there a watchdog timer?
+The RP2XXX port even has a Watchdog.The watchdog is a countdown timer that can restart parts of the chip if it reaches zero. This can be used to restart the processor if software gets stuck in an infinite loop. The programmer must periodically write a value to the watchdog to stop it from reaching zero.
```
Comment (mcauser) [inline docs/rp2/quickref.rst:232]: Watchdog.The -> Watchdog. The

File: docs/rp2/tutorial/intro.rst [hunk 10]
```diff
@@ -1,6 +1,247 @@
 .. _rp2_intro:
 
-Getting started with MicroPython on the RP2xxx
-==============================================
+1.Getting started with MicroPython on the RP2xxx
+================================================
+
+The Pico board is the most popular in the series and we will be talking about it in general for all the RP2xxx boards.This tutorial will guide you through setting up MicroPython, connecting to the network and communicating with the Internet, using the hardware peripherals, and controlling some external components.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:6]: boards.This -> boards. This

File: docs/rp2/tutorial/intro.rst [hunk 11]
```diff
@@ -1,6 +1,247 @@
...
+
+1.1. The MicroPython Environment
+---------------------------------
+
+Python is the fastest way to get started with embedded software on RP2xxx ports.Without compromising on approachability, MicroPython can be used for embedded systems development at a professional level.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:11]: ports.Without -> ports. Without

File: docs/rp2/tutorial/intro.rst [hunk 12]
```diff
@@ -1,6 +1,247 @@
...
+The arihmtic performance of the is quite accurate as the accelerated floating point libraries in RP2040’s on-board ROM storage are used automatically by the Python code.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:12]: arihmtic -> arithmetic

File: docs/rp2/tutorial/intro.rst [hunk 13]
```diff
@@ -1,6 +1,247 @@
...
+
+We also have ease of working on existing MicroPython projects as they can ported easily because of the standard *machine* module being used for most of the on-chip hardware.The second processor core is exposed through the *_thread* module.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:14]: hardware.The -> hardware. The

File: docs/rp2/tutorial/intro.rst [hunk 14]
```diff
@@ -1,6 +1,247 @@
...
+
+1.3. Powering the board
+------------------------
+
+You can power your board with the help of a USB connector if the board has it otherwise it needs to be powered directly.
+
+1.4. Getting the firmware
+--------------------------
+
+To get MicroPython the easiest way, download the pre-built release binary from the `Pico Getting Started pages https://www.raspberrypi.org/documentation/rp2040/getting-started/`_. If you can’t or don’t want to use the pre-built release or you want to build your own MicroPython firmware binary like  developing your own C module for MicroPython — you can follow section 1.4 to  get the source code for MicroPython,
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:36]: A few double-spaces

File: docs/rp2/tutorial/intro.rst [hunk 15]
```diff
@@ -1,6 +1,247 @@
...
+
+1.4. Getting the firmware
+--------------------------
+
+To get MicroPython the easiest way, download the pre-built release binary from the `Pico Getting Started pages https://www.raspberrypi.org/documentation/rp2040/getting-started/`_. If you can’t or don’t want to use the pre-built release or you want to build your own MicroPython firmware binary like  developing your own C module for MicroPython — you can follow section 1.4 to  get the source code for MicroPython, 
+
+1.5. Installing MicroPython on Raspberry Pi Pico
+-------------------------------------------------
+
+Raspberry Pi Pico has a BOOTSEL mode for programming firmware over the USB port.First step is to ensure that the board  is not plugged into any source of power: disconnect the micro USB cable if plugged in, and disconnect any other wires that might be providing power to the board, e.g. through the VSYS or VBUS pin. Now hold down the BOOTSEL button, and plug in the micro USB cable which must ne connected to a computer.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:41]: port.First -> port. First

File: docs/rp2/tutorial/intro.rst [hunk 16]
```diff
@@ -1,6 +1,247 @@
...
+
+Drag the MicroPython *firmware.uf2* file onto a drive called the RPI-RP2 which pops up. This programs the MicroPython firmware(the UF2 File) onto the flash memory on your Raspberry Pi Pico.
+After finishing,the board reboots and boots into MicroPython, waiting for further instructions.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:44]: finishing,the -> finishing, the

File: docs/rp2/tutorial/intro.rst [hunk 17]
```diff
@@ -1,6 +1,247 @@
...
+Check for the source code in a new directory called micropython. The MicroPython repository also contains pointers (submodules) to specific versions of libraries it needs to run on a particular board, like the SDK in the case of RP2040. We need to explicitly fetch these too:
+
+.. code-block:: bash
+
+	$ cd micropython
+	$ git submodule update --init -- lib/pico-sdk lib/tinyusb
+
+.. note:: 
+
+	The following instructions assume that you are using a Raspberry Pi Pico. Some details may differ if you are building firmware for a different RP2040-based board. The board vendor should detail any extra 	    steps needed to build firmware for that particular board. These instructions are generic, but might differ at places like putting the default serial port on different pins,or including extra modules to drive that board’s hardware.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:86]:
```
extra 	    steps -> extra steps
```

File: docs/rp2/tutorial/intro.rst [hunk 18]
```diff
@@ -1,6 +1,247 @@
...
+	$ make -C mpy-cross
+
+We can now build the port we need for RP2040, that is, the version of MicroPython that has specific support for our chip.
+
+.. code-block:: bash
+
+	$ cd portal/rp2
+	$ make
+
+After this,check for a directory called *build* ( *ports/rp2/build* relative to the micropython directory), which contains the new firmware binaries. The most important ones are:
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:108]: this,check -> this, check

File: docs/rp2/tutorial/intro.rst [hunk 19]
```diff
@@ -1,6 +1,247 @@
...
+There are two ways to connect to this REPL, so you can communicate with the MicroPython firmware on your board: over USB, and over the UART serial port on Raspberry Pi Pico GPIOs.
+
+2.1. Connecting from a Raspberry Pi over USB
+---------------------------------------------
+
+The MicroPython firmware is equipped with a virtual USB serial port which is accessed through the micro USB connector on Raspberry Pi Pico. Your computer should notice this serial port and list it as a character device, most likely */dev/ttyACM0.*
+
+.. note::
+
+	You can run ls /dev/tty* to list your serial ports. There may be quite a few, but MicroPython’s USB serial will start with /dev/ttyACM. If in doubt, unplug the micro USB connector and see which 		one disappears. If you don’t see anything, you can try rebooting your Raspberry Pi.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:140]:
```
which 		one -> which one
```

File: docs/rp2/tutorial/intro.rst [hunk 20]
```diff
@@ -1,6 +1,247 @@
...
+This is a prompt. MicroPython wants you to type something in, and tell it what to do, similar to any other command prompt on any OS.Press CTRL-D and a message will pop up:
+
+.. topic:: Message
+
+	MPY: soft reboot
+	MicroPython v1.13-422-g904433073 on 2021-01-19; Raspberry Pi Pico with RP2040
+	Type "help()" for more information.
+	>>>
+
+This key combination tells MicroPython to reboot. You can do this at any time. When it reboots, MicroPython will printout a message saying exactly what firmware version it is running, and when it was built. Your version number will be different from the one shown here.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:168]: printout -> print out

File: docs/rp2/tutorial/intro.rst [hunk 21]
```diff
@@ -1,6 +1,247 @@
...
+
+2.2. Connecting from a Raspberry Pi Pico using UART
+---------------------------------------------------
+
+.. warning::
+
+	REPL over UART is disabled by default.
+
+The MicroPython port for RP2040 does not expose REPL over a UART port by default. However this default can bechanged in the ports/rp2/mpconfigport.h source file.To use the REPL over UART first build MicroPython (see section 1.6 for details).
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:177]: file.To -> file. To

File: docs/rp2/tutorial/intro.rst [hunk 22]
```diff
@@ -1,6 +1,247 @@
...
+        Raspberry Pi                Raspberry Pi Pico
+        =================           ==================
+        GND                         GND
+        GPIO15(UART_RX0)            GPIO0(UART_TX)
+        GPIO14(UART_TX0)            GPIO1(UART_RX)
+        =================           ==================
+
+.. warning:: 
+
+	RX matches to TX, and TX matches to RX. You mustn’t connect the two opposite TX pins together, or the two RX pins. This is because MicroPython needs to listen on the channel that the Raspberry Pi 		transmits on, and vice versa.
```
Comment (mcauser) [inline docs/rp2/tutorial/intro.rst:238]:
```
Raspberry Pi 		transmits -> Raspberry Pi transmits
```

**Review by NitiKaur — COMMENTED** [review]
@mcauser  I have made the changes too. I think it is all set for a merge

**Comment by mcauser** [comment]
Hey @NitiKaur, looking good, only a few left:
docs/esp32/quickref.rst:609 `sources.hey` -> `sources. They`
docs/rp2/quickref.rst:224 `Watchdog.The` -> `Watchdog. The`
docs/rp2/tutorial/intro.rst:6 `boards.This` -> `boards. This`
docs/rp2/tutorial/intro.rst:105 `this,check` -> `this, check`

**Comment by kadamski** [comment]
@NitiKaur It would make sense to state what you are doing in your commit messages. If you think about it, the "Update intro.rst" doesn't say much about the commit alone. For example, there are 5 different intro.rst files in the tree, so it would make sense to add "rp2: " prefix. Since it is documentation, the refix could be "docs/rp2: ". Then "update" is almost as generic as "change". Doesn't say what you change there. Maybe you fix something important, maybe you only fix some typos, or grammar or white spaces.

It also makes a lot of sense to group the changes logically - create a separate commit with simple fixes like whitespace fixes, typos, etc - changes that are obvious. Then have a separate commit with more serious changes that might be debatable or add completely new content and may require some deeper review.

Those are just advises, I'm not sure if this is required for this PR to be requested.

**Comment by NitiKaur** [comment]
@kadamski Thank you for your suggestion. I will take care of it the nest time :)

**Comment by NitiKaur** [comment]
Creating a new PR with the changes in ADC, PWM, Timer and watchdog class as discussed  in the new PR

**Pull request closed without merging**